src/aten/sycl/BatchNormKernels.cpp

```diff
@@ -0,0 +1,3965 @@
+  auto sg = item.get_sub_group();
+  uint32_t lane_id = sg.get_local_linear_id();
+  uint32_t sg_id = sg.get_group_linear_id();
+
+  // dynamic get SIMD width result in big performance drop
+  // uint32_t SIMD = sg.get_local_range()[0];
+#pragma unroll
+  for (int i = 1; i < SIMD; i <<= 1) {
+    val = bin_op(val, static_cast<accscalar_t>(sg.shuffle_down(val, i)));
+  }
+  if (sub_group_num == 1) {
+    if (lane_id == 0) {
+      local_data[0] = val;
+    }
+    item.barrier(sycl_local_fence);
+    val = local_data[0];
+
+    return;
+  }
+
+  // reduce internal each subgroup, each subgroup will generate one result
+  // there are WGroupSize/subGroupSize elements after this step
+  if (lane_id == 0) {
+    local_data[sg_id] = val;
+  }
+  item.barrier(sycl_local_fence);
+
+  // use one subgroup to reduce WGroupSize/subGroupSize elements
+  // into the final result
+  if (sg_id == 0) {
+    val = init;
+    if (lane_id < sub_group_num) {
+      val = accscalar_t(local_data[lane_id]);
+    }
+    for (int i = lane_id + SIMD; i < sub_group_num; i += SIMD) {
+      val = bin_op(val, static_cast<accscalar_t>(local_data[i]));
+    }
+#pragma unroll
+    for (int i = 1; i < SIMD; i <<= 1) {
+      val = bin_op(val, static_cast<accscalar_t>(sg.shuffle_down(val, i)));
+      if (i >= ((sub_group_num + 1) >> 1))
+        break;
+    }
+
+    // the 0th WI (the 0th WI in the 0th sub_group) generate the final
+    // result
+    if (lane_id == 0) {
+      local_data[0] = val;
+    }
+  }
+
+  item.barrier(sycl_local_fence);
+  val = local_data[0];
+}
+
+template <
+    int SIMD,
+    typename scalar_t,
+    typename item_t,
+    typename Op,
+    typename PTA,
+    typename local_shared_t>
+scalar_t plane_reduce(
+    item_t item,
+    Op grad_op,
+    PTA tensor,
+    int plane,
+    int sub_group_num,
+    const local_shared_t& shared) {
+  // first the reductions each thread does separately
+  scalar_t sum_value = 0;
+  for (int batch = item.get_local_id(0); batch < tensor.size(0);
+       batch += item.get_local_range(0)) {
+    for (int x = item.get_local_id(1); x < tensor.size(2);
+         x += item.get_local_range(1)) {
+      auto res = grad_op(batch, plane, x);
+      sum_value += res;
+    }
+  }
+  // group_reduce<SIMD, scalar_t>(
+  //     item,
+  //     sub_group_num,
+  //     sum_value,
+  //     scalar_t(0),
+  //     shared,
+  //     [](scalar_t a, scalar_t b) { return a + b; });
+  // if (item.get_local_linear_id() == 0) {
+  //   shared[0] = sum_value;
+  // }
+  // item.barrier(sycl_local_fence);
+  // // Everyone picks it up, should be broadcast into the whole grad_input
+  // return shared[0];
+
+  shared[item.get_local_linear_id()] = sum_value;
+  item.barrier(sycl_local_fence);
+  scalar_t output = 0;
+  for (int i = 0; i < SIMD * sub_group_num; i++) {
+    output += shared[i];
+  }
+  item.barrier(sycl_local_fence);
+  return output;
+}
+
+template <typename scalar_t>
+int inline get_nhwc_suggest_vec_size(
```

Comment:
Let us use channels_last? Although there is no alignment in stock CUDA, we can follow PyTorch declaration to use channels last. The semantics is broader than nhwc.